### PR TITLE
Setup CI

### PR DESCRIPTION
**Pull request by SimonHeybrock**

### CODE REVIEW

**Comment by nvaytet** [comment]
Is there a danger of this rapidly getting out of date/sync with the scipp setup if we change something e.g. in the `scipp/docs/conf.py` file and we don't change it here also?

**Comment by SimonHeybrock** [comment]
Not really, unless your changes to `scipp/docs/conf.py` are incompatiable with the templates here... but in that case the scipp CI should fail.

**Comment by SimonHeybrock** [comment]
There is one problem though (which I think can and should be fixed in the future): The templates appear to be relying on hard-coded paths in many cases. Should probably use parameters instead.

**Comment by nvaytet** [comment]
Why do you need `main` and `release` pipelines here? You could just test everything in the PR, including deployment, to some dummy conda package in the scipp channel?

**Comment by nvaytet** [comment]
> There is one problem though (which I think can and should be fixed in the future): The templates appear to be relying on hard-coded paths in many cases. Should probably use parameters instead.

Yes, an example of this is your latest failed build, which would also fail if we tried to use this for the `ess` repo, since the package name does not begin with `scipp`

**Comment by SimonHeybrock** [comment]
I think I *do* need all pipelines (PR, Main, Release), since I want to test all those templates (though the may be unifyable?). But you are right that I want to run all those pipelines on PRs, i.e., I need to change the triggers?

**Comment by nvaytet** [comment]
Ok yes, changing the triggers to be the same as PR should be fine.
Not sure if we want to run them again once we merge to main, maybe it's not necessary.

**Comment by SimonHeybrock** [comment]
Can I pass a list of params, e.g., `deploy: [false, true]` so I do not need separate yml files but still test wit both params?

**Comment by nvaytet** [comment]
when you have `deploy: true` you are testing all the same as `false`, but you execute additional steps.
so i think it's would be fine to just have `deploy: true` on the PR.

However, you do slightly different things when doing `release: true`, so you may want to test those 2 cases instead.
I don't really see a good way of doing this cleanly. You don't want to restart the entire thing (with code quality, and c++ tests), ideally you would just want to re-run the deploy step a second time with `release: true`

**Comment by SimonHeybrock** [comment]
I don't mind restarting everything, basically just setup a build matrix from those params.

**Comment by nvaytet** [comment]
What happens if you do something like
```yaml
trigger:
  - main

pr: none

stages:
  - template: stages.yml
    parameters:
      verbose: ${{ parameters.verbose }}
      conda_channels: '--channel scipp --channel ess-dmsc'
      gh_pages_subdirectory: 'scippneutron'
      config:
        linux:
          py_versions: ['3.7']
          conda_env: 'scippneutron-developer.yml'
        osx:
          py_versions: ['3.7']
          conda_env: 'scippneutron-developer.yml'
        windows:
          py_versions: ['3.7']
          conda_env: 'scippneutron-developer-no-mantid.yml'
      deploy: true
      conda_label: 'dev'

  - template: stages.yml
    parameters:
      verbose: ${{ parameters.verbose }}
      conda_channels: '--channel scipp --channel ess-dmsc'
      gh_pages_subdirectory: 'scippneutron'
      config:
        linux:
          py_versions: ['3.7']
          conda_env: 'scippneutron-developer.yml'
        osx:
          py_versions: ['3.7']
          conda_env: 'scippneutron-developer.yml'
        windows:
          py_versions: ['3.7']
          conda_env: 'scippneutron-developer-no-mantid.yml'
      deploy: true
      conda_label: 'dev'
      release: true
```
?

**Comment by nvaytet** [comment]
Or even
```yaml
parameters:
- name: verbose
  displayName: Verbose
  type: boolean
  default: false
- name: scipp_channel
  type: string
  default: scipp
- name: releases
  type: object
  default: [true, false]

trigger:
  branches:
    exclude:
      - '*'

pr:
  autoCancel: true
  branches:
    include:
      - '*'
stages:
  - ${{ each r in parameters.releases }}:
    - template: stages.yml
      parameters:
        verbose: ${{ parameters.verbose }}
        conda_channels: '--channel scipp --channel ess-dmsc'
        gh_pages_subdirectory: 'scippneutron'
        config:
          linux:
            py_versions: ['3.7']
            conda_env: 'scippneutron-developer.yml'
          osx:
            py_versions: ['3.7']
            conda_env: 'scippneutron-developer.yml'
          windows:
            py_versions: ['3.7']
            conda_env: 'scippneutron-developer-no-mantid.yml'
        deploy: true
        conda_label: 'dev'
        release: ${{ r }}
```

**Comment by SimonHeybrock** [comment]
@nvaytet Decided to keep things simple for now: Not testing deployment, this can be added later if required, and what we have here should be better than nothing. Please have a look.

File: conda/conda_build_config.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+python:
+  - 3.7
```
Comment (nvaytet) [inline conda/conda_build_config.yaml:2]: I'm not sure we need this file if we always pass the version to the `conda build` command?

**Review by nvaytet — COMMENTED** [review]
One question. Otherwise, looks fantastic.

File: conda/conda_build_config.yaml (hunk 1 above)
Comment (nvaytet) [reply conda/conda_build_config.yaml:2]: So i checked locally and it does override the version, so i think you can remove that file.
And maybe you can also remove `python` from the `conda/meta.yaml` file?

Comment (SimonHeybrock) [reply conda/conda_build_config.yaml:2]: "maybe"? Can I or not?

Comment (nvaytet) [reply conda/conda_build_config.yaml]: "Maybe" here meant "can you give it a go and we can see if it still works as we think"?

File: test/.azure-pipelines/pull_request.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    include:
+      - '*'
+
+extends:
+  template: ../../stages.yml
+  parameters:
+    verbose: ${{ parameters.verbose }}
+    config:
+      linux:
+        py_versions: ['3.7']
```
Comment (nvaytet) [inline test/.azure-pipelines/pull_request.yml:24]: Do we want to see if it also works for multiple python versions?

Comment (SimonHeybrock) [reply test/.azure-pipelines/pull_request.yml:24]: Wouldn't that be a test of the `yml` syntax rather than our templates?

**Review by nvaytet — APPROVED** [review]

**Pull request merged**